diff --git a/PKGBUILD b/PKGBUILD
index f64d315..52a0ca6 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -8,7 +8,7 @@ pkgver=42.3
 pkgrel=1
 pkgdesc="A window manager for GNOME"
 url="https://gitlab.gnome.org/GNOME/mutter"
-arch=(x86_64)
+arch=(x86_64 aarch64)
 license=(GPL)
 depends=(dconf gobject-introspection-runtime gsettings-desktop-schemas
          libcanberra startup-notification zenity libsm gnome-desktop 
@@ -19,8 +19,10 @@ makedepends=(gobject-introspection git egl-wayland meson xorg-server
 checkdepends=(xorg-server-xvfb wireplumber python-dbusmock)
 options=(debug)
 _commit=1903356b45c6c884a0451580f32494ff1288656d  # tags/42.3^0
-source=("git+https://gitlab.gnome.org/GNOME/mutter.git#commit=$_commit")
-sha256sums=('SKIP')
+source=("git+https://gitlab.gnome.org/GNOME/mutter.git#commit=$_commit"
+  "https://gitlab.com/hrdl/pinenote-shared/-/raw/main/patches/mutter/0001-Add-META_CONNECTOR_TYPE_DPI.patch")
+sha256sums=('SKIP'
+  'SKIP')
 
 pkgver() {
   cd mutter
@@ -29,6 +31,7 @@ pkgver() {
 
 prepare() {
   cd mutter
+  patch -p1 < "$srcdir/0001-Add-META_CONNECTOR_TYPE_DPI.patch"
 }
 
 build() {
@@ -43,25 +46,5 @@ build() {
   meson compile -C build
 }
 
-_check() (
-  mkdir -p -m 700 "${XDG_RUNTIME_DIR:=$PWD/runtime-dir}"
-  glib-compile-schemas "${GSETTINGS_SCHEMA_DIR:=$PWD/build/data}"
-  export XDG_RUNTIME_DIR GSETTINGS_SCHEMA_DIR
-
-  pipewire &
-  _p1=$!
-
-  wireplumber &
-  _p2=$!
-
-  trap "kill $_p1 $_p2; wait" EXIT
-
-  meson test -C build --print-errorlogs -t 3
-)
-
-check() {
-  dbus-run-session xvfb-run -s '-nolisten local +iglx -noreset' \
-    bash -c "$(declare -f _check); _check"
-}
 
 _pick() {
